engine/access/subscription/blocks_watcher.go

```diff
@@ -0,0 +1,209 @@
+	if startBlockID == h.RootBlockID ||
+		// Note: there is a corner case when rootBlockHeight == 0:
+		// since the default value of an uint64 is 0, when checking if startHeight matches the root block
+		// we also need to check that startBlockID is unset, otherwise we may incorrectly set the start height
+		// for non-matching startBlockIDs.
+		(startHeight == h.RootHeight && startBlockID == flow.ZeroID) {
+		return h.RootHeight + 1, nil
+	}
+
+	var header *flow.Header
+	var err error
+	// invalid or missing block IDs will result in an error
+	if startBlockID != flow.ZeroID {
+		header, err = h.headers.ByBlockID(startBlockID)
+		if err != nil {
+			return 0, rpc.ConvertStorageError(fmt.Errorf("could not get header for block %v: %w", startBlockID, err))
+		}
+
+		if blockStatus == flow.BlockStatusFinalized {
+			return header.Height, nil
+		}
+	}
+
+	// heights that have not been indexed yet will result in an error
+	if startHeight > 0 {
+		if startHeight < h.RootHeight {
+			return 0, status.Errorf(codes.InvalidArgument, "start height must be greater than or equal to the root height %d", h.RootHeight)
+		}
+
+		header, err = h.headers.ByHeight(startHeight)
+		if err != nil {
+			return 0, rpc.ConvertStorageError(fmt.Errorf("could not get header for height %d: %w", startHeight, err))
+		}
+
+		if blockStatus == flow.BlockStatusFinalized {
+			return header.Height, nil
+		}
+	}
+
+	lastSealed, err := h.state.Sealed().Head()
+	if err != nil {
+		return 0, status.Errorf(codes.Internal, "could not get latest sealed block: %v", err)
+	}
+
+	// this checking if start block is sealed
+	if header != nil {
+		if header.Height > lastSealed.Height {
+			return 0, status.Errorf(codes.InvalidArgument, "provided start block must be sealed. Latest sealed block %v with the height %d", lastSealed.ID(), lastSealed.Height)
+		}
+
+		return header.Height, nil
+	}
+
+	// if no start block was provided, use the latest sealed block
+	return lastSealed.Height, nil
+}
+
+// GetHighestHeight returns the highest height based on the specified block status. Only flow.BlockStatusFinalized and flow.BlockStatusSealed allowed.
+func (h *BlocksWatcher) GetHighestHeight(blockStatus flow.BlockStatus) (uint64, error) {
+	switch blockStatus {
+	case flow.BlockStatusFinalized:
+		return h.finalizedHighestHeight.Value(), nil
+	case flow.BlockStatusSealed:
+		return h.sealedHighestHeight.Value(), nil
+	case flow.BlockStatusUnknown:
+		return 0, status.Errorf(codes.InvalidArgument, "could not get highest height for block with unknown status")
+	}
+
+	return 0, status.Errorf(codes.InvalidArgument, "could not get highest height for invalid status")
+}
+
+// SetFinalizedHighestHeight sets the highest finalized block height.
+func (h *BlocksWatcher) SetFinalizedHighestHeight(height uint64) bool {
+	return h.finalizedHighestHeight.Set(height)
+}
+
+// SetSealedHighestHeight sets the highest sealed block height.
+func (h *BlocksWatcher) SetSealedHighestHeight(height uint64) bool {
```

Comment:
Any reason for it to be public? I find this a bit confusing, I would prefer to eliminate these methods and use private counters in finalization handler.